gcloud-java-resourcemanager/src/main/java/com/google/gcloud/resourcemanager/testing/LocalResourceManagerHelper.java

```diff
@@ -0,0 +1,593 @@
+
+/**
+ * Utility to create a local Resource Manager mock for testing.
+ *
+ * <p>The mock runs in a separate thread, listening for HTTP requests on the local machine at an
+ * ephemeral port.
+ */
+@SuppressWarnings("restriction")
+public class LocalResourceManagerHelper {
+  private static final JsonFactory jsonFactory =
+      new com.google.api.client.json.jackson.JacksonFactory();
+  private static final Random PROJECT_NUMBER_GENERATOR = new Random();
+  private static final String VERSION = "v1beta1";
+
+  // see https://cloud.google.com/resource-manager/reference/rest/v1beta1/projects
+  private static final Set<Character> PERMISSIBLE_PROJECT_NAME_PUNCTUATION =
+      ImmutableSet.of('-', '\'', '"', ' ', '!');
+
+  private HttpServer server;
+  private final ConcurrentHashMap<String, Project> projects = new ConcurrentHashMap<>();
+  private final int port;
+
+  static class Response {
+    private final int code;
+    private final String body;
+
+    Response(int code, String body) {
+      this.code = code;
+      this.body = body;
+    }
+
+    int code() {
+      return code;
+    }
+
+    String body() {
+      return body;
+    }
+  }
+
+  enum Error {
+    ALREADY_EXISTS(409, "global", "alreadyExists", "ALREADY_EXISTS"),
+    PERMISSION_DENIED(403, "global", "forbidden", "PERMISSION_DENIED"),
+    // change failed precondition error code to 412 when #440 is fixed
+    FAILED_PRECONDITION(400, "global", "failedPrecondition", "FAILED_PRECONDITION"),
+    // change invalid argument error code to 412 when #440 is fixed
+    INVALID_ARGUMENT(400, "global", "badRequest", "INVALID_ARGUMENT"),
+    BAD_REQUEST(400, "global", "badRequest", "BAD_REQUEST"),
+    INTERNAL_ERROR(500, "global", "internalError", "INTERNAL_ERROR");
+
+    private final int code;
+    private final String domain;
+    private final String reason;
+    private final String status;
+
+    Error(int code, String domain, String reason, String status) {
+      this.code = code;
+      this.domain = domain;
+      this.reason = reason;
+      this.status = status;
+    }
+
+    Response response(String message) {
+      try {
+        return new Response(code, toJson(message));
+      } catch (IOException e) {
+        return Error.INTERNAL_ERROR.response("Error when generating JSON error response");
+      }
+    }
+
+    private String toJson(String message) throws IOException {
+      Map<String, Object> args = new HashMap<>();
+      Map<String, Object> errors = new HashMap<>();
+      errors.put("domain", domain);
+      errors.put("message", message);
+      errors.put("reason", reason);
+      args.put("errors", ImmutableList.of(errors));
+      args.put("code", code);
+      args.put("message", message);
+      args.put("status", status);
+      return jsonFactory.toString(ImmutableMap.of("error", args));
+    }
+  }
+
+  private class RequestHandler implements HttpHandler {
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
```

Comment:
I suggest you remove the `throws IOException` and make sure you handle all exception that may be thrown by the individual calls. all failures should be translated into a json error response if possible.

You will probably need to catch the exception from the OutputStream.write though (and log it as ERROR).